app/frontend/testnet/testnet-actions.ts

```diff
@@ -0,0 +1,167 @@
+      interval = null
+      callback(...args)
+    }, time)
+  }
+}
+
+type SetStateFn = (newState: Partial<State>) => void
+type GetStateFn = () => State
+export default ({setState, getState}: {setState: SetStateFn; getState: GetStateFn}) => {
+
+  const loadWallet = async(state, {cryptoProviderType, walletSecretDef}) => {
+    const wallet = await TestnetWallet(
+      {walletSecretDef}
+    )
+    const accountInfo = await wallet.getAccountStatus(walletSecretDef.rootSecret.pubkeyHex)
+    const delegationHistory = await wallet.getDelegationHistory()
+    const validStakepools = await getValidStakePools()
+    const displayStakingPage = true
+    setState({
+      validStakepools,
+      walletIsLoaded: true,
+      ...accountInfo,
+      delegationHistory,
+      displayStakingPage,
+    })
+  }
+
+  const submitTransaction = (state) => {
+
+  }
+
+  const changeDelegation = async (state) => {
+    console.log('changing delegation')
+  }
+
+  const reloadWalletInfo = () => {
+    return null
+  }
+
+  const calculateDelegatedPercent = (pools) => {
+    return pools.map((pool) => pool.percent).reduce((x, y) => x + y, 0)
+  }
+
+  // DUMMY
+  const fetchPoolInfo = (poolId) => {
+    return poolId === 'adalite-stake-pool-id'
+      && {
+        name: 'AdaLite Stake Pool',
+      }
+  }
+
+  // DUMMY
+  const calculateDelegationFee = () => {
+    const state = getState()
+    setState({
+      delegationFee: state.stakePools.length,
+    })
+    setState({
+      calculatingDelegationFee: false,
+    })
+  }
+
+  const debouncedCalculateDelegationFee = debounceEvent(calculateDelegationFee, 2000)
+
+  // DUMMY
+  const validateDelegationAndCalculateDelegationFee = () => {
+    const state = getState()
+    const isValid =
+      state.stakePools.every((pool) => pool.valid && pool.percent > 0) &&
+      calculateDelegatedPercent(state.stakePools) === 100
+
+    setState({ isDelegationValid: isValid })
+    if (isValid) {
+      setState({ calculatingDelegationFee: true })
+      debouncedCalculateDelegationFee()
+    } else {
+      setState({ delegationFee: 0 })
+    }
+  }
+
+  const updateStakePoolId = (state, e) => {
+    const poolId = e.target.value
+    setState({
+      stakePools: state.stakePools.map((pool, i) => {
+        return Object.assign(
+          {},
+          pool,
+          i === parseInt(e.target.name, 10)
+            ? {
+              id: e.target.value,
+              valid: state.validStakepools.has(poolId),
+              ...fetchPoolInfo(poolId),
+            }
+            : {}
+        )
+      }),
+    })
+    validateDelegationAndCalculateDelegationFee()
+  }
+
+  const updateStakePoolPercent = (state, e) => {
+    const index = parseInt(e.target.name, 10)
+    const delegatedPercent = calculateDelegatedPercent(state.stakePools)
+    const oldPercent = state.stakePools
+      .map((pool, i) => (i === index ? pool.percent : 0))
+      .reduce((x, y) => x + y, 0)
+    const nextPercent = Math.min(
```

Comment:
What is going on here? Why do we do `math.min` at all and just not set the value from the event?